### ISSUE TITLE
1000+ unrendered spheres causes a memory spike, crashing Chrome tab

### ISSUE THREAD

**Issue by dbryanbio** [3dmol/3Dmol.js#367]
Process memory spikes to > 4G when adding 1000+ spheres, then subsides when rendering them. 
Crash a Chrome tab with 1022 unrendered spheres.

To reproduce:
Rename the attached files to 3Dmol.js (May 2017 release) and sphere-crash-repro.html, then view sphere-crash-repro.html in Chrome on Windows.  Click "Add Spheres" three times to crash the Chrome tab.  If you open up the Chrome Task Manager with Shift-Escape, you can watch the memory increasing and decreasing when you add spheres and render them.

A successful way to add many spheres is to add and render in chunks of 500.
My environment is Windows 10, Dell XPS 15, 32G Ram.

[3Dmol.txt](https://github.com/3dmol/3Dmol.js/files/1907635/3Dmol.txt)
[sphere-crash-repro.txt](https://github.com/3dmol/3Dmol.js/files/1907636/sphere-crash-repro.txt)

**Comment by dkoes** [comment]
This is somewhat by design.  Each call to addSphere is creating a new shape.  Shapes don't truncate their buffers until render time so that additional primitives can be added to the shape with low memory overhead.

Can you try making `manySpheres` be a GLShape instead of a list?  
```
manySpheres = glviewer.addShape({color: 'red'});
...
manySpheres.addSphere({center: {x: x, y: y, z: z}, radius: 1, color: 'red'});
```
This workaround only works if you want all the spheres to share the same material (color).

I think it would be reasonable for the GLViewer shape functions to truncate the buffers since most people building up a shape object would be calling the GLShape functions directly.  I'll try to get that pushed soon, but can you verify that the above workaround fixes the problem on your end?

**Comment by dbryanbio** [comment]
Hi David,
Thanks for your reply.  This definitely works in the test program and
performs much better than my other workaround of re-rendering every 800
spheres.

It may not work for our application though.  We want to be able to hover on
individual spheres; they create different tooltips based on their data.  I
didn't see a way to get a hold of the individual sphere that triggered the
hover event within a shape.

My boss will like the improved performance though...maybe we'll have to
rethink our tooltips!

Best,
Dan Bryan

**Comment by dkoes** [comment]
I've pushed a fix (ee13fe0a7ac5bbb3f585ca85ea3c5e66f66a34d5) so spheres created through the GLViewer helper functions are more memory efficient.  If you are treating a bunch of shape primitives as one shape with shared material properties, creating a single GLShape and adding to that is the way to go.

However, if all you want spheres with individual tooltips, your best bet is to create a molecule with a bunch of "atoms" rendered with a fixed radius.  As you can imagine, 3Dmol.js is optimized for rendering atoms.  They can all have their own hover callback (http://3dmol.csb.pitt.edu/tests/auto/generate_test.cgi?test=glviewer_js_setHoverable) and can have different, explicitly set, radii and colors.

**Comment by dbryanbio** [comment]
David, thanks for your help!

Our spheres are actually transparent halos around atom groups, so actually
using atoms wouldn't work.  But thank you again for the tips and the fix.
3Dmol is great to work with.
Cheers!
